### DISCUSSION TITLE
[Proposal] Case fall through should be allowed

### DISCUSSION THREAD

**Discussion by dotnetchris** [dotnet/csharplang#603] [General]
C# should really allow proper case fall through. There is no reason it needs to derivate from C with how switch behaves. 

Perhaps an alternative would be similar to 

`switch unchecked(foo)` and this would allow proper case fall through without changing anything existing.

Having located https://github.com/dotnet/roslyn/issues/168 it already seems like this is supported and that something just specifically bars us from using it normally. Whatever disallows us to use case fall through should be turned off or have an opt-out.

Original thread: https://github.com/dotnet/roslyn/issues/1426

**Comment by HaloFour** [comment]
> There is no reason it needs to derivate from C with how switch behaves.

C allows for case fallthrough and it's a massive source of bugs, which is why it was decided that C# would disallow it.

**Reply by Kilian-Bosse** [reply]
how about a fallthrough; keyword as the opposite of break; ?
0 chance for unintentional use while being far more clear and less prone to confusion (after all, we've heard the cries about never using "goto", which doesn't apply at all in this case) than the current messy option.

**Reply by tannergooding** [reply]
`fallthrough` still has plenty of risk, such as across refactorings where control flow then goes to the wrong place.

`goto case ...` is explicit and avoids this problem altogether.

**Reply by timcassell** [reply]
> we've heard the cries about never using "goto"

And those cries are wrong when it comes to C#. `goto` is perfectly safe and a reasonable tool to use in C#.

**Reply by tannergooding** [reply]
It's "safer" because C# doesn't allow arbitrary `goto` and you can't violate definite assignment; however you still don't want to do it because it can mess with control flow analysis especially for IL and often will lead to worse overall codegen in many scenarios.

There's some isolated cases where you can use it to your benefit, but more often its still better to simply avoid it altogether and write code with "normal" control flow. `goto case ...` in a switch is one of the exceptions because of its limitations and the general representation of `switch` as a jump table in the kind of scenarios you'd typically use such a "fallthrough".

**Comment by jnm2** [comment]
Explicitly writing <code>goto case *nextcase;*</code> is the way to go. It's not like you're stopping the compiler from optimizing that out versus fallthrough.

**Comment by DavidArno** [comment]
@jnm2,
> Explicitly writing `goto case nextcase;` is the way to go...

That's a matter of opinion. 

Writing structured code that doesn't require implicit, or explicit, "goto next case" is the real way to go. It's worth watching Jules May's "If considered harmful" talk for a detailed description of why test/goto's like this probably account for 95% of all bugs in existing code.

**Comment by jnm2** [comment]
That's actually a good point.

**Comment by DavidArno** [comment]
Thank you. I make them occasionally 😉

**Comment by Richiban** [comment]
Funnily enough, I actually didn't realise that C# *didn't* allow case fallthrough. The fact that it doesn't, however, begs the question "Why the he** is the `break` keyword required then?".

**Comment by HaloFour** [comment]
@Richiban 

C# does allow for case fallthrough, but only under the specific scenario of a switch-section with an empty statement-list.  And the `break` keyword isn't required, you can use any control statement that would result in leaving the `switch` block, or explicitly jumping to a different `case` label via `goto`.

**Comment by Richiban** [comment]
@HaloFour Sure thing. You can, of course, use other control statements. But why can't I write this?

```csharp
void M(int i)
{
	switch (i)
	{
		case 1: Console.WriteLine("one");
		case 2: Console.WriteLine("two");
		default: Console.WriteLine("Some other number");
	}
}
```

**Comment by HaloFour** [comment]
@Richiban 

Probably because most people (at the time) would expect that syntax would fallthrough.  Forcing the developer to exit the switch-section but not automatically doing so keeps the expected syntax and avoids bugs.  If C# were being designed today the argument might have been different.

**Comment by DavidArno** [comment]
@HaloFour 
> C# does allow for case fallthrough...

As [Eric Lippert explains here](https://stackoverflow.com/a/3049601/7122), C# does not support cases with empty statement lists. That is  a C-language way of thinking about C#'s switches. In reality, C# treats it as multiple case labels  for a section of the `switch` statement. (Caveat: this may have changed with the v7 changes to the language).

@Richiban,
As  Eric explains in that same link, the way C# enforces no case fall-through is by mandating that the end of the statement list be unreachable, so that list must end in `goto`, `break`, `return` etc. They didn't need to do it that way, so it was purely a design decision.

**Comment by Richiban** [comment]
@HaloFour @DavidArno Ahh, okay. Thanks for the info.

Since I've never done any C or other similar level programming I'm not able to think in the historical context in which C# was first designed.

Still, surely there's the possibility that this requirement can be removed now? After all, it's not a breaking change if the `break` statement suddenly becomes redundant. I'm assuming that C# has plans to add a match statement to the language; does that mean that switch will effectively have been replaced? Why not just upgrade switch to support a more modern syntax?

**Comment by HaloFour** [comment]
@Richiban 

> Still, surely there's the possibility that this requirement can be removed now?

Theoretically.  It couldn't possibly break (pun unintended) any existing source.  But is that really worthwhile?  Sure, you eliminate six characters (and likely a newline) per switch-section, which isn't a massive improvement.  And that still doesn't change the perceived expectation, especially for anyone coming from C, C++, Objective-C or Java, and there are significantly more developers using those languages than any of the "modern" languages.

> I'm assuming that C# has plans to add a match statement to the language; does that mean that switch will effectively have been replaced?

Still under design but based on the various discussions on the Roslyn repo it seems that such a construct would be strictly for expressions.  A section could not contain multiple statements nor could it produce a value of `void`.  The LDM didn't seem particularly receptive to requests to support either.

> Why not just upgrade switch to support a more modern syntax?

Probably because then the language would have two identical ways of doing something with very slightly different syntax and behavior.  It seems a little silly to me to chase "modern" constructs as that is a continuously moving target.  Eventually even Apple Swift will seem passé and some people will clamor that it (and every other language) copy the syntax from the new hotness.

**Comment by tannergooding** [comment]
I prefer the explicit `goto case` syntax over something like a `fallthrough` keyword as well. It helps prevent bugs that would otherwise exist from code refactorings.

**Comment by Richiban** [comment]
@HaloFour 

> It seems a little silly to me to chase "modern" constructs as that is a continuously moving target

You're right; "modern" was absolutely the wrong word here. I think I use it simply because I am very unsympathetic to the "That's the way it's always been done" argument. What I really meant was "non-C" programming. 

There are plenty of programmers who (whether it's through discovery of their own or through teaching) simply don't use 'jump' keywords such as `break`, `continue` or `goto`. To these people the use of `break` in `switch` is both puzzling and ugly. That's why I was looking forward to `match` so much but to hear you say 

>such a construct would be strictly for expressions

is, I'm sorry, completely nuts. Why on earth would they do that? It means that: 

> the language would have two identical ways of doing something

with the only difference being whether you're writing a statement or an expression!?

**Comment by svick** [comment]
@Richiban Is having `if` as well as the ternary operator also "completely nuts"?

**Comment by Richiban** [comment]
@svick 

> Is having if as well as the ternary operator also "completely nuts"?

Yes.

I understand why it's there though; C had a very clear delineation between statements and expressions. This is a viewpoint that's gradually falling out of favour though; in any language (that I can think of) that doesn't have its roots in C everything is an expression and the following:

```
val a = if (b) "yes" else "no"
```

is perfectly valid and understandable.

Again, I totally understand why C# has these legacy constructs considering its history--I just don't understand why the language design team would voluntarily choose _to continue following_ these patterns.

**Comment by jnm2** [comment]
> I just don't understand why the language design team would voluntarily choose to continue following these patterns.

I know that consistency with the feel and theory of existing C# is a consideration with some priority. That would cause the team to voluntarily continue following existing patterns.

**Comment by DavidArno** [comment]
@svick,

To add to @Richiban's reply, not only is it nuts to have `if` and the ternary operator, the latter is ugly and difficult to read compared with using `if/else` for expressions, especially if more than one test expression is needed:
```cs
var result = expression1 ? 1 : expression2 ? 2 : 3;
var result = if (expression1) 1 else if (expression2) 2 else 3;
```

And I do not hold with the idea that just because there is an existing way to do something, we shouldn't introduce newer, better ways to do the same thing. Roslyn analysers could even be used to guide folk into using the newer syntax (that approach worked brilliantly with R# offering to rewrite C# 5 code to take advantage of C# 6 features for example). 

Not having `match` support "void expressions" (ie statements) would indeed be nuts.

**Comment by svick** [comment]
> in any language (that I can think of) that doesn't have its roots in C everything is an expression

There are many languages that don't have C roots and yet don't use the "everything is an expression" approach, including Pascal, Python or Go. In fact, I'd say "everything is an expression" is specific to functional languages and languages heavily influenced by functional programming.

In any case, C# is not such a language and wishing that it was is not likely to end with a good design, I think.

> I just don't understand why the language design team would voluntarily choose to *continue following* these patterns

You're making it sound as if it was an option to have one reasonable way of writing a `switch`/`match`, that would work as an expression and as a statement, but it's not.

The options are:

1. Add a version of `switch` that works as an expression and keep the existing `switch` statement.
2. Add a completely new `match` that works as an expression and keep the existing `switch` statement.
3. Add a completely new `match` that works as a statement or an expression and keep the existing `switch` statement.

In every case, there are two different ways to write the same thing.

Option 1 sounds close to your goal of having only one way of writing a `switch`, but it still wouldn't be the same, because `switch` was not designed to work as an expression.

Option 3 could not remove the `switch` statement, so you still would have two different ways to write a `switch`.

**Comment by Richiban** [comment]
@svick 

> In every case, there are two different ways to write the same thing.

Number 1 doesn't sound like two ways of writing the same thing.

> because switch was not designed to work as an expression

I'm afraid I don't understand this argument. I don't see why `switch` can't just get some minor upgrades to support expressions. Let's assume the following syntax:

```csharp
// Statement-based switch
void M(int i)
{
    switch(i)
    {
        case 0: Console.WriteLine("zero");
        case 1: Console.WriteLine("one");
        default: Console.WriteLine("some other number");
    }
}

// Expression-based switch
void M2(int i)
{
    Console.WriteLine(
        switch (i)
        {
            case 0: "zero";
            case 1: "one";
            default: "some other number";
        }
    );
}
```

I think it's entirely possible to make one syntactic construct that works as both, and I think there's no reason that can't be switch.

**Comment by HaloFour** [comment]
The same was suggested of `if` and that was rejected by the LDM.  I don't think they have much interest in making C# a pervasively expression-based language like functional languages.

The bonus here is that if the team is considering a new syntax for expression-based `switch`/`match` that they can ditch the existing baggage.

**Comment by svick** [comment]
> I think it's entirely possible to make one syntactic construct that works as both

I don't think it could really be one syntactic construct. In the statement form, each `case` has a sequence of one or more statements; in the expression form, each `case` has a single expression followed by a semicolon (so it looks like a statement). Those are not the same thing.

So you end up with with two versions of `switch`, that kinda-sorta look similar, but are not the same and where the expression form is inconsistent with other expressions in the language (the semicolons).

Which is why I think this kind of almost-consistency between expressions and statements should not have high priority when designing the `switch`/`match` expression.

**Comment by Richiban** [comment]
Oops, you're right about the semicolons. I wasn't paying enough attention when that was written.

I've amended it such:

```csharp
void M2(int i)
{
    Console.WriteLine(
        switch (i)
        {
            case 0: "zero"
            case 1: "one"
            default: "some other number"
        }
    );
}
```

That should be perfectly unambiguous when it comes to nested switch statements since the switch body must be surrounded by braces.

I'd love to know what the LDM is thinking at the moment, since I've not seen any actual details on these pages, just that "they are working on a match statement". I, for one, would favour a syntax similar to lambdas:

```
void M3(int i)
{
    Console.WriteLine(
        match i (
            case 0 => "zero",
            case 1 => "one",
            case * => "some other number"
        )
    );
}
```

And I think the idea of statements / expressions is already being blurred with other proposals, such as sequence expressions:

```

var john = (
    var firstName = "John";
    var secondName = "Adams";

    new Person(firstName, secondName)
);
```

**Comment by tannergooding** [comment]
I don't think the only thing to consider is whether or not it is unambiguous in C#.

In my opinion, it also should be straightforward to understand for someone who is coming from a non C# background or for someone who is working in multiple languages to mentally parse/understand the code without having to remember special language rules.

For example, given the following code:
```
switch (i)
{
    case 0: "zero";
    case 1: "one";
    default: "some other number";
}
```

It means one thing in C/C++ and is disallowed in C#. If we allowed it (keeping current C# behavior of no-case fall through). It now has one meaning in C/C++ and another meaning in C#. If I am working in a mixed-language code-base, I now have to make more context switches and possible have to think harder to ensure I am properly parsing through the code when reading it. It may also make it harder for those coming from another language to learn C# the first time, which could potentially slow adoption of the language by new people.

One of the biggest strengths of C#, I believe, is that it is very familiar to C/C++ users (so it is relatively easy to pickup), but it also forces the user to be explicit in cases that are a common source of bugs for C/C++ users.

**Comment by svick** [comment]
@tannergooding Technically, that's not valid C or C++, because of missing semicolons. But that's a small distinction to make and it could indeed cause confusion.

**Comment by tannergooding** [comment]
@svick, I just fixed it to be valid 😄

**Comment by CyrusNajmabadi** [comment]
>  I don't think they have much interest in making C# a pervasively expression-based language like functional languages.

I actually think there's a lot of interest for this.  However, that doesn't mean we'd want to reuse existing statement forms when making expression forms.  For one thing, it can just be harder to easily grok what's going on when you're scanning the code.

For that reason, i think we're strongly behind some sort of expression-based pattern matching construct.  However, that doesn't necessarily mean we'd try to reuse the switch/case keywords in it.  I personally would like something pretty lightweight.  'switches' are already pretty heavyweight , and that feels like it would be unfortunate to me in an expression context.  

At the very least, i can't see why we'd keep 'case' (or any other keyword) when it can so easily be elided in expression contexts.

**Comment by HaloFour** [comment]
@CyrusNajmabadi 

>  However, that doesn't mean we'd want to reuse existing statement forms when making expression forms. 

This is what I mean; that the team doesn't seem interested in making current statements into expressions.  Similarly, there doesn't seem to be a lot of interest in making the new expression forms supported as statements.  On Roslyn there were several requests that expression `match`/`switch` would support `void` expressions so that you could use it in place of the current `switch` statement and that idea seemed to get shot down.  Granted, I'm sure nothing there is set in stone yet.

**Comment by lachbaer** [comment]
My idea to making the `break` optional without sacrificing read ambiguities is this:
```cs
switch (i) break    // additional `break` introduces the new (optional) way
{
    case 0: "zero";
    case 1: "one";
    case 2:  // warning, empty case breaks and does not fall through
    case 3: "three";
    default: "some other number";
}
```

**Comment by Richiban** [comment]
@CyrusNajmabadi I'm glad to hear that the new match statement is going to be lighter weight; switch/case really was too verbose.

I'm going to make this point only once more (I'm aware that we're hijacking the thread somewhat with this discussion) but I _really_ recommend against disallowing `match` in a statement context. We've already talked about how it makes things feel more consistent (switch is now just a legacy construct, like `delegate` instead of a lambda) but the argument that the same construct should be used for both is actually getting stronger, especially since the `do` LINQ query terminator now actually has a champion:  https://github.com/dotnet/csharplang/issues/101, which is basically giving LINQ expressions the ability to act like statements!

A little anecdote that explains why I'm hot on this topic right now: I'm currently teaching my SO how to program (she's done none at all before) and this is something that comes up regularly that she finds really confusing. I think she's largely got the difference between an expression and a statement (although she really backed me into a corner when it came to methods that return data but you ignore the result. "Is an expression or a statement?!" "Uhh... both?"), but I still see her trying to assign a `foreach` loop to a variable or use the ternary operator bare, as a statement.

Anyway, I think that if a new match statement is introduced to the language it's going to have a bunch of advantages and a lighter syntax so we should move away from switch as a result, not create this artificial divide for no reason other than to not have essentially deprecated `switch`.

**Comment by jnm2** [comment]
@DavidArno I'm noticing the [If Considered Harmful](https://www.youtube.com/watch?v=z43bmaMwagI) talk affecting the way I factor my code. Thanks for that. The talk was slow but worthwhile.

**Comment by AustinBryan** [comment]
You could also do

```csharp
switch(myInt)
{
   case 0: case0Func(); break;
   case 1: case0Func();
       case1Func();
       break;
    default: defaultFunc(); break;
}

```

Which will make them more readable too.

**Comment by DavidArno** [comment]
@AustinBryan,

Since it is now proposed that [switch expressions allow the use of statement expressions](https://github.com/dotnet/csharplang/issues/1054#issuecomment-391479240), I'd argue that the most readable way to express that would be:
```cs
myInt switch
{
    0 => case0Func();
    1 => case1Func();
    _ => defaultFunc();
};

...

void case1Func()
{
    case0Func();
    ....
}
```
No fall-through required and if `case0Func` and `case1Func` need calling for the `1` case, then make that the responsibility of `case1Func`, not the switch.

**Comment by dotnetchris** [comment]
@DavidArno the immediate thought i have is that defeats the true purposes of case fall through, the few but very specialized optimizations you can make in tight loops. In a tight loop that's going to explode the stack with thousands/millions of method calls. 

Unless the compiler is smart enough to inline so intelligently it creates the proper CLR level case fall through with no stack usage at all. But that seems like a very tall order.

**Comment by harrisse** [comment]
Sorry if someone has already proposed this already, but what about using `continue` in `switch` for case fallthrough? I don't have much of an issue with `goto case` in constant expressions, but the addition of pattern matching in `switch` makes it hard to `goto`.

As an example, say `OurFoo : IFoo`, `MyFoo : OurFoo`, `YourFoo : OurFoo`:
```
public void ProcessFoo(IFoo foo)
{
    switch (foo)
    {
        case MyFoo myFoo:
            // Process MyFoo
            continue;
        case YourFoo yourFoo:
            // Process YourFoo
            continue;
        case OurFoo ourFoo:
            // Process OurFoo
            return;
    }
}
```
This would allow `YourFoo` and `MyFoo` to be processed both like our respective types, but also like `OurFoo`.

**Comment by CyrusNajmabadi** [comment]
This already has meaning in the language.  The swithc may be in a continuable construct (like a for-loop).  Such things are common.  This would make the meaning very confusing.

**Comment by harrisse** [comment]
@CyrusNajmabadi I think I understand your point, but a switch can also be in a breakable construct and we allow break in switch statements so to me that feels like the language is getting more consistent (switch [or case?] becomes both breakable and continuable -- more like other constructs like for and while).

**Comment by harrisse** [comment]
Though I do understand it could be backwards-incompatible.

**Comment by YairHalberstadt** [comment]
I literally wrote code yesterday that continued out of a switch statement. I would imagine it would be very backwards incompatible.

**Comment by yaakov-h** [comment]
FWIW, Swift has a `fallthrough` keyword last I checked...

**Comment by The-Futurist** [comment]
We could consider some of the features that were in the more powerful PL/I `select` [statement.](https://www.ibm.com/support/knowledgecenter/SSQ2R2_14.1.0/com.ibm.ent.pl1.zos.doc/topics/lsh-select.html) 

This was much less primitive than the feeble C switch, allowing expressions in places that C only allows constants.

Might there be scope for making a `case` support expressions rather than just constants?

In fact such an extension would have automatically provided the "type" pattern support as just being a special case of a more general expression based match.

```cs
switch (variable.GetType())
{
case typeof(string)
        ;
        break;
case typeof(List<string>)
        ;
        break;
}
```
